src/Workspaces/Core/Portable/Workspace/Solution/TextLoader.cs

```diff
@@ -29,83 +34,107 @@ public abstract class TextLoader
         /// <exception cref="IOException" />
         /// <exception cref="InvalidDataException"/>
         /// <exception cref="OperationCanceledException"/>
-        public abstract Task<TextAndVersion> LoadTextAndVersionAsync(Workspace workspace, DocumentId documentId, CancellationToken cancellationToken);
+        public virtual Task<TextAndVersion> LoadTextAndVersionAsync(CancellationToken cancellationToken)
+        {
+#pragma warning disable CS0618 // Type or member is obsolete
+            if (ImmutableInterlocked.GetOrAdd(
+                ref s_isObsoleteLoadTextAndVersionAsyncOverriden,
+                GetType(),
+                new Func<Workspace, DocumentId, CancellationToken, Task<TextAndVersion>>(LoadTextAndVersionAsync).Method.DeclaringType != typeof(TextLoader)))
```

Comment:
>Method [](http://example.com/codeflow?start=114&length=6)

Are we relying on delegate object to give us a final override? Is this a documented contract?

Reply:
Checked with CLR ream -- this is correct.

Reply:
> Checked with CLR ream -- this is correct.

Is this documented?

Reply:
It is. The compiler emits:

```
ldvirtftn instance void TextLoader::LoadTextAndVersionAsync()
newobj instance System.Func<...>::.ctor(object, native int)
```

`ldvirtftn` loads the address of the implementation of the given virtual method. So the method pointer passed to the delegate is to the implementation of the method.